util/dedup/dedupCache.go

```diff
@@ -208,7 +221,7 @@ func (c *txSaltedCache) innerCheck(msg []byte) (*crypto.Digest, bool) {
 
 // CheckAndPut adds msg into a cache if not found
 // returns a hashing key used for insertion if the message not found.
-func (c *txSaltedCache) CheckAndPut(msg []byte) (*crypto.Digest, bool) {
+func (c *SaltedCache) CheckAndPut(msg []byte) (*crypto.Digest, bool) {
```

Comment:
We still see some duplicate transactions going through with very short intervals in between them.

If we want to remove them should we do another check right before writing? Right now we are only doing it in case of salt having changed but if two (or more) messages pass the RLock simulatenously (returning hash misses) and get to line 235 where they acquire write Lock, it looks to me like, they will both be processed as if they are not duplicates. 

If we remove the conditional `salt != c.curSalt` we would prevent this case but introduce an additional unconditional hash on all messages that get past the first check.

Reply:
yeah, this is @cce's introduction of RW Lock.